proof/crefine/AARCH64/IsolatedThreadAction.thy

```diff
@@ -1624,25 +1637,180 @@ lemma threadGet_isolatable:
   apply (erule obj_atE')
-  apply (clarsimp simp: projectKOs
-                        partial_overwrite_def put_tcb_state_regs_def
+  apply (clarsimp simp: partial_overwrite_def put_tcb_state_regs_def
                   cong: if_cong)
   apply (simp add: projectKO_opt_tcb v split: if_split)
   done
 
+lemma typ_at_partial_overwrite_id2:
+  "\<lbrakk> \<forall>x. tcb_at' (idx x) s \<rbrakk>
+    \<Longrightarrow> typ_at' P p (ksPSpace_update (partial_overwrite idx f) s)
+             = typ_at' P p s"
+  apply (frule dom_partial_overwrite[where tsrs=f])
+  apply (simp add: typ_at'_def ko_wp_at'_def obj_at'_def ps_clear_def partial_overwrite_def
+                    split: if_split)
+  apply clarsimp
+  apply (drule_tac x=x in spec)
+  apply (clarsimp simp: put_tcb_state_regs_def objBits_simps
+                        project_inject)
+  done
+
+lemma tcb_at_partial_overwrite:
+  "\<forall>x. tcb_at' (idx x) s \<Longrightarrow>
+   tcb_at' p (ksPSpace_update (partial_overwrite idx f) s) = tcb_at' p s"
+  by (simp add: tcb_at_typ_at' typ_at_partial_overwrite_id2)
+
+
+definition asUserFPU :: "machine_word \<Rightarrow> (fpu_state \<Rightarrow> ('a \<times> fpu_state)) \<Rightarrow> 'a kernel" where
+  "asUserFPU tptr f \<equiv> do
+        fpu \<leftarrow> threadGet (fpu_state \<circ> atcbContextGet o tcbArch) tptr;
+        threadSet (\<lambda>tcb. tcb \<lparr> tcbArch := atcbContextSet (UserContext (snd (f fpu)) (user_regs (atcbContext (tcbArch tcb))))
+                                           (tcbArch tcb)\<rparr>) tptr;
+        return (fst (f fpu))
+   od"
+
+lemma asUser_asUserFPU_set:
+  "asUser t (setFPUState fc) = asUserFPU t (\<lambda>_. ((), fc))"
+  unfolding asUser_def asUserFPU_def setFPUState_def
+  apply (clarsimp simp: select_f_returns threadGet_stateAssert_gets bind_assoc exec_gets comp_def
+                intro!: ext bind_apply_cong[where f="stateAssert P L" for P L, OF refl]
+                 dest!: fst_stateAssertD)
+  apply (clarsimp simp: threadSet_det simpler_modify_def)
+  apply (rule kernel_state.unfold_congs; simp?)
+  apply (rule ext)
+  apply (clarsimp simp: obj_at'_def atcbContextSet_def thread_fetch_def atcbContextGet_def)
+  done
+
+lemma asUser_asUserFPU_get:
+  "asUser t getFPUState = asUserFPU t (\<lambda>fc. (fc, fc))"
+  unfolding asUser_def asUserFPU_def getFPUState_def
+  apply (subst threadSet_stateAssert_modify)+
+  apply (clarsimp simp: select_f_returns threadGet_stateAssert_gets bind_assoc exec_gets comp_def
+                intro!: ext bind_apply_cong[where f="stateAssert P L" for P L, OF refl]
+                 dest!: fst_stateAssertD)
+  apply (clarsimp simp: simpler_modify_def return_def bind_def)
+  apply (clarsimp simp: obj_at'_def atcbContextSet_def thread_fetch_def atcbContextGet_def thread_replace_def)
+  done
+
+lemma put_tcb_state_regs_tcb_twice[simp]:
+  "put_tcb_state_regs_tcb tsr (put_tcb_state_regs_tcb tsr' tcb)
+    = put_tcb_state_regs_tcb tsr tcb"
```

Comment:
indentation: the point of left-wrap operator style is to not have to put extra indentation in, `=` should be under `put`